activerecord/lib/active_record/store.rb

```diff
@@ -290,9 +290,11 @@ def self.as_indifferent_hash(obj)
           def as_regular_hash(obj)
             case obj
             when ActiveSupport::HashWithIndifferentAccess
-              obj.to_h
+              obj.to_hash
             when Hash
-              obj
+              obj.transform_values { |value| value.is_a?(Hash) || value.is_a?(Array) ? as_regular_hash(value) : value }
+            when Array
+              obj.map { |value| value.is_a?(Hash) || value.is_a?(Array) ? as_regular_hash(value) : value }
```

Comment:
We don't need that part of the change. Just the `to_h -> to_hash` is enough.

Reply:
@byroot Thank you for reviewing. If I am understanding it correctly, we should only call `HWIA#to_hash` and drop nested handling for `Hash` & `Array`. i.e
```ruby
def as_regular_hash(obj)
  case obj
  when ActiveSupport::HashWithIndifferentAccess
    obj.to_hash
  when Hash
     obj            
  else
    {}
   end
end
```

Reply:
Yes. `to_hash` will take care of the recursive behavior. If the top level object is already a regular hash, there is no point recursively checking the object graph.

Reply:
Actually, even simpler: `obj ? obj.to_hash : {}`, since `Hash#to_hash` returns self.

Reply:
👍

Reply:
@byroot I opted for `obj.respond_to?(:to_hash) ? obj.to_hash : {}` as we have a test that fails when the passed object doesn't implement `#to_hash`. 

TEST "convert store attributes from any format other than Hash or HashWithIndifferentAccess losing the data" => `activerecord/test/cases/store_test.rb:208`